File: go/BinaryTree/1448_count-good-nodes-in-binary-tree/1448_count-good-nodes-in-binary-tree.go
package countgoodnodesinbinarytree

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func goodNodes(root *TreeNode) int {
	maxValue := root.Val
	return goodNodesUtil(root, maxValue)
}

func goodNodesUtil(node *TreeNode, maxValue int) int {
	var result int
	if node == nil {
		return 0
	}
	if node.Val >= maxValue {
		result = 1
	} else {
		result = 0
	}
	maxValue = max(node.Val, maxValue)
	result += goodNodesUtil(node.Left, maxValue)
	result += goodNodesUtil(node.Right, maxValue)
	return result
}

func max(i, j int) int {
	if i > j {
		return i
	}
	return j
}

File: c/1448-Count-Good-Nodes-in-Binary-Tree.c
/*
Given a binary tree root, a node X in the tree is named good if in the path from root to X there are no nodes with a value greater than X.
Return the number of good nodes in the binary tree.
Time: O(n)
Space: O(log(h)) Where h is the height of the tree
*/

int nbGood(struct TreeNode* root, int m) {
    if (root==NULL)
        return 0;
    if (root->val >= m)
        return 1+nbGood(root->left, root->val)+nbGood(root->right, root->val);
    return nbGood(root->left, m)+nbGood(root->right, m);
}

int goodNodes(struct TreeNode* root){
    return nbGood(root, INT_MIN);
}
